File: services/ai/app/schemas.py
"""Pydantic models mirroring packages/shared/src/schemas.ts (the Zod contract)."""
from typing import List, Optional, Literal, Dict, Any
from pydantic import BaseModel, Field

Point = Dict[str, float]


class Room(BaseModel):
    label: str
    rawLabel: Optional[str] = None
    type: str
    polygon: List[List[float]]
    centroid: List[float]
    area: float
    confidence: float = Field(ge=0, le=1)
    source: Literal["cv", "manual"] = "cv"
    reviewed: bool = False


class ZoneGeometry(BaseModel):
    kind: Literal["point", "line", "polygon"]
    coords: List[List[float]]


class Zone(BaseModel):
    type: str
    geometry: ZoneGeometry
    confidence: float = Field(ge=0, le=1)
    source: Literal["cv", "manual"] = "cv"


class Detection(BaseModel):
    cls: str = Field(alias="class")
    bbox: List[float]
    confidence: float

    class Config:
        populate_by_name = True


class Placement(BaseModel):
    deviceCode: str
    label: Optional[str] = None
    position: Point
    rotation: float = 0
    scale: float = 1
    locked: bool = False
    hidden: bool = False
    source: Literal["ai", "manual"] = "ai"
    reviewed: bool = False
    rationale: Optional[str] = None
    confidence: Optional[float] = None
    props: Dict[str, Any] = {}
    meta: Dict[str, Any] = {}
    zIndex: int = 0


class QcRejection(BaseModel):
    deviceCode: str = "-"
    reason: str
    confidence: Optional[float] = None
    nearSpace: Optional[str] = None


class AnalysisQcSummary(BaseModel):
    detectedSpaces: int = 0
    acceptedSpaces: int = 0
    rejectedSpaces: int = 0
    rawPlacements: int = 0
    acceptedPlacements: int = 0
    rejectedPlacements: int = 0
    rejections: List[QcRejection] = []


class AnalysisResult(BaseModel):
    floorId: Optional[str] = None
    image: Dict[str, int]
    rooms: List[Room] = []
    zones: List[Zone] = []
    detections: List[Detection] = []
    placements: List[Placement] = []
    confidence: float = 0
    provider: Literal["cv", "llm_fallback"] = "cv"
    warnings: List[str] = []
    qcSummary: Optional[AnalysisQcSummary] = None
    rawRoomCount: Optional[int] = None


class QcOverrides(BaseModel):
    """Admin-tunable quality-control limits forwarded by the worker. All optional;
    unset fields fall back to the defaults in rules/quality.py."""
    maxRoomsPerFloor: Optional[int] = None
    maxDevicesPerFloor: Optional[int] = None
    maxDevicesPerRoom: Optional[int] = None
    minRoomConfidence: Optional[float] = None
    minDeviceConfidence: Optional[float] = None


class AnalyzeRequest(BaseModel):
    imageUrl: Optional[str] = None
    imageB64: Optional[str] = None
    floorId: Optional[str] = None
    units: str = "m"
    provider: Literal["cv", "llm_fallback"] = "cv"
    fallbackProvider: str = "disabled"
    qc: Optional[QcOverrides] = None


class SuggestRequest(BaseModel):
    rooms: List[Room]
    zones: List[Zone] = []


class IngestPage(BaseModel):
    name: str
    width: int
    height: int
    b64: str


class IngestResponse(BaseModel):
    pages: List[IngestPage]
    warnings: List[str] = []

File: packages/shared/src/schemas.ts
import { z } from 'zod';
import { ROOM_TYPES, ZONE_TYPES } from './space-types';

export const pointSchema = z.object({ x: z.number(), y: z.number() });

export const roomSchema = z.object({
  id: z.string().optional(),
  label: z.string(),
  rawLabel: z.string().optional(),
  type: z.enum(ROOM_TYPES),
  polygon: z.array(z.array(z.number())),
  centroid: z.tuple([z.number(), z.number()]),
  area: z.number(),
  confidence: z.number().min(0).max(1),
  source: z.enum(['cv', 'manual']),
  reviewed: z.boolean().optional(),
  reviewStatus: z.enum(['ai_detected', 'rejected', 'accepted', 'user_corrected']).optional(),
  aiType: z.enum(ROOM_TYPES).nullable().optional(),
  aiConfidence: z.number().nullable().optional(),
  rejectionReason: z.string().nullable().optional(),
  meta: z.record(z.unknown()).optional(),
});

/** Body for manual space creation / room edits from the editor's Spaces panel. */
export const createRoomSchema = z.object({
  type: z.enum(ROOM_TYPES),
  label: z.string().optional(),
  polygon: z.array(z.array(z.number())).optional(),
  centroid: z.tuple([z.number(), z.number()]).optional(),
  area: z.number().optional(),
});

export const updateRoomSchema = z.object({
  type: z.enum(ROOM_TYPES).optional(),
  label: z.string().optional(),
  reviewStatus: z.enum(['ai_detected', 'rejected', 'accepted', 'user_corrected']).optional(),
  centroid: z.tuple([z.number(), z.number()]).optional(),
  polygon: z.array(z.array(z.number())).optional(),
  // Manual marking: flag a space as double-height so ceiling devices (Wi-Fi/speakers)
  // relocate to the nearest normal-ceiling space per the customer rule.
  doubleHeight: z.boolean().optional(),
}).refine((b) => Object.keys(b).length > 0, { message: 'No fields to update' });

export const zoneSchema = z.object({
  id: z.string().optional(),
  type: z.enum(ZONE_TYPES),
  geometry: z.object({
    kind: z.enum(['point', 'line', 'polygon']),
    coords: z.array(z.array(z.number())),
  }),
  confidence: z.number().min(0).max(1),
  source: z.enum(['cv', 'manual']),
});

export const placementSchema = z.object({
  id: z.string().optional(),
  deviceCode: z.string(),
  label: z.string().optional(),
  position: pointSchema,
  rotation: z.number().default(0),
  scale: z.number().positive().default(1),
  layerId: z.string().optional(),
  groupId: z.string().optional(),
  locked: z.boolean().default(false),
  hidden: z.boolean().default(false),
  source: z.enum(['ai', 'manual']).default('manual'),
  reviewed: z.boolean().default(false),
  rationale: z.string().optional(),
  confidence: z.number().min(0).max(1).optional(),
  props: z.record(z.unknown()).default({}),
  meta: z.record(z.unknown()).optional(),
  zIndex: z.number().default(0),
});

export const analysisResultSchema = z.object({
  floorId: z.string().optional(),
  image: z.object({ width: z.number(), height: z.number() }),
  rooms: z.array(roomSchema),
  zones: z.array(zoneSchema),
  detections: z.array(z.object({ class: z.string(), bbox: z.array(z.number()), confidence: z.number() })),
  placements: z.array(placementSchema),
  confidence: z.number().min(0).max(1),
  provider: z.enum(['cv', 'llm_fallback']),
  warnings: z.array(z.string()),
  qcSummary: z.object({
    detectedSpaces: z.number(),
    acceptedSpaces: z.number(),
    rejectedSpaces: z.number(),
    rawPlacements: z.number(),
    acceptedPlacements: z.number(),
    rejectedPlacements: z.number(),
    roomBasedPlacements: z.number().optional(),
    zoneBasedPlacements: z.number().optional(),
    perimeterBasedPlacements: z.number().optional(),
    consistent: z.boolean().optional(),
    summary: z.string().optional(),
    rejections: z.array(z.object({
      deviceCode: z.string(),
      reason: z.string(),
      confidence: z.number().optional(),
      nearSpace: z.string().optional(),
    })),
  }).optional(),
  rawRoomCount: z.number().optional(),
  scale: z.object({
    metersPerPixel: z.number(),
    confidence: z.number(),
    source: z.string(),
  }).optional(),
});

// ── API DTO schemas ──
export const registerSchema = z.object({
  tenantName: z.string().min(2),
  name: z.string().min(2),
  email: z.string().email(),
  password: z.string().min(8),
});
export const loginSchema = z.object({ email: z.string().email(), password: z.string().min(1) });
export const createProjectSchema = z.object({
  name: z.string().min(1),
  code: z.string().optional(),
  units: z.enum(['m', 'ft']).default('m'),
  client: z.object({
    name: z.string().optional(), contact: z.string().optional(),
    phone: z.string().optional(), email: z.string().email().optional(), address: z.string().optional(),
  }).optional(),
});
export const batchPlacementSchema = z.object({
  upserts: z.array(placementSchema).default([]),
  deletes: z.array(z.string()).default([]),
});

export type AnalysisResultInput = z.infer<typeof analysisResultSchema>;
